builtin/reset.c

```diff
@@ -174,88 +174,6 @@ static void update_index_from_diff(struct diff_queue_struct *q,
 	}
 }
 
-static int pathspec_needs_expanded_index(const struct pathspec *pathspec)
```

Comment:
It can help to reference `4d1cfc1351 (reset: make --mixed sparse-aware, 2021-11-29)` in your commit message as the commit that introduced the method.

Reply:
> Add your compatibility tests in t1092 first. Make them all be test_expect_success and never make a change that breaks them.

I'm not so sure how can this work. I think it needs a few changes to make some of the tests successful along the way. For example, without marking `the_repository->settings.command_requires_full_index = 0;`, the index expands unexpectedly and it fails the `ensure_not_expanded` tests.

So I guess some of the tests, for example, `ensure_not_expanded` need to be `test_expect_failure` at the first commit.